modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1215,6 +1220,8 @@ private void distributedExchange() throws IgniteCheckedException {
 
         changeWalModeIfNeeded();
 
+        cctx.time().schedule(this::onDistributedExchangeTimeout, cctx.gridConfig().getExchangeHardTimeout(), -1);
```

Comment:
Is this supposed behavior that the returned CancelledTask is not re-scheduled in case of the coordinator node left the cluster? Let's assume that the coordinator node left the cluster for some reason and therefore, a new node will be elected. New coordinator has to request single messages from all participated nodes and so, there is a chance onDistributedExchangeTimeout() method will be called. I mean that time required for exchange on new coordinator is less than getExchangeHardTimeout(). Perhaps, it's not an issue at all.

Reply:
It is intended behavior. ExchangeHardTimeout is maximum amount of time that user can afford to spend on exchange. If this timeout is reached, we should do everything to finish exchange ASAP.

Reply:
Thank you, Ilya. Got it.